cdap-master/src/main/java/co/cask/cdap/data/tools/HBaseTableExporter.java

```diff
@@ -0,0 +1,222 @@
+
+  public HBaseTableExporter() throws Exception {
+    this.hConf = HBaseConfiguration.create();
+    this.cConf = CConfiguration.create();
+    Injector injector = createInjector(cConf, hConf);
+    this.txClient = injector.getInstance(TransactionSystemClient.class);
+    this.txService = injector.getInstance(TransactionService.class);
+    this.zkClientService = injector.getInstance(ZKClientService.class);
+
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        try {
+          HBaseTableExporter.this.stop();
+        } catch (Throwable e) {
+          LOG.error("Failed to stop the tool.", e);
+        }
+      }
+    });
+
+  }
+
+  private static Injector createInjector(CConfiguration cConf, Configuration hConf) {
+    return Guice.createInjector(
+      new ConfigModule(cConf, hConf),
+      new IOModule(),
+      new ZKClientModule(),
+      new KafkaClientModule(),
+      new LocationRuntimeModule().getDistributedModules(),
+      new DiscoveryRuntimeModule().getDistributedModules(),
+      new DataFabricModules().getDistributedModules(),
+      new DataSetsModules().getDistributedModules(),
+      new MetricsClientRuntimeModule().getDistributedModules(),
+      new LoggingModules().getDistributedModules(),
+      new ExploreClientModule(),
+      new ViewAdminModules().getDistributedModules(),
+      new StreamAdminModules().getDistributedModules(),
+      new NotificationFeedServiceRuntimeModule().getDistributedModules()
+    );
+  }
+
+  /**
+   * Sets up the actual MapReduce job.
+   * @param tx The transaction which needs to be passed to the Scan instance. This transaction is be used by
+   *           coprocessors to filter out the data corresonding to the invalid transactions .
+   * @param tableName Name of the table which need to be exported as HFiles.
+   * @return the configured job
+   * @throws IOException
+   */
+  public Job createSubmittableJob(Transaction tx, String tableName) throws IOException {
+
+    Job job = Job.getInstance(hConf, "HBaseTableExporter");
+
+    job.setJarByClass(HBaseTableExporter.class);
+    Scan scan = new Scan();
+    scan.setCacheBlocks(false);
+    // Set the transaction attribute for the scan.
+    scan.setAttribute(TxConstants.TX_OPERATION_ATTRIBUTE_KEY, new TransactionCodec().encode(tx));
+    job.setNumReduceTasks(0);
+
+    TableMapReduceUtil.initTableMapperJob(tableName, scan, Import.KeyValueImporter.class, null, null, job);
+
+    FileSystem fs = FileSystem.get(hConf);
+    Random rand = new Random();
+    Path root = new Path(fs.getWorkingDirectory(), "hbasetableexporter");
+    fs.mkdirs(root);
+    while (true) {
+      bulkloadDir = new Path(root, "" + rand.nextLong());
+      if (!fs.exists(bulkloadDir)) {
+        break;
+      }
+    }
+
+    HFileOutputFormat2.setOutputPath(job, bulkloadDir);
+    HTable hTable = new HTable(hConf, tableName);
+    HFileOutputFormat2.configureIncrementalLoad(job, hTable);
```

Comment:
so you are reading from a table T1 and you want export its data to HFiles that can be loaded into Table T2. But you configure the output of the map.reduce to be optimized for loading into the T1. (For example, this sets the number of reducers to the number of regions in the source table). 

What happens if you load these HFiles into a fresh table on the new cluster? That table by default has only one region. Have you tested this with a source table that has more than one region?

I think you either need to 
- configure the job to optimize output for the destination table, or
- ensure that the destination table is created with the same regions as the source. For that you would have to capture the partitions file somehow, and use that to pre-split the new table into the same regions. 

The second option is probably the better one because it reflects the state of the source table more accurately.